File: Algorithm/Other/leetcode/73. Set Matrix Zeroes/Claude/Set-Matrix-Zeroes.py
# # Python コーディング問題解析・実装・検証

# ## 1. 問題分析結果

# ### 競技プログラミング視点
# - **制約分析**: m×n行列、1≤m,n≤200 → 最大40,000要素、O(mn)が必須
# - **最速手法**: in-place操作でO(1)空間、CPythonの組み込み関数活用
# - **メモリ最小化**: 追加リスト作成を回避、原配列直接操作

# ### 業務開発視点
# - **型安全設計**: List[List[int]]の厳密な型ヒント + mypy/pylance対応
# - **エラーハンドリング**: 入力検証、境界条件チェック
# - **可読性**: docstring、型ヒント、明確な変数名による自己文書化

# ### Python特有分析
# - **データ構造選択**: list の in-place 操作が最適
# - **標準ライブラリ活用**: 基本的な操作のため不要
# - **CPython最適化**: range(), enumerate(), built-in関数の活用

# ## 2. アルゴリズム比較表

# |アプローチ|時間計算量|空間計算量|Python実装コスト|可読性|標準ライブラリ活用|CPython最適化|備考|
# |-----|-----|-----|-----------|---|-----------------|----------|--|
# |追加配列使用|O(mn)|O(m+n)|低|★★★|不要|適|理解しやすい|
# |第一行列使用|O(mn)|O(1)|中|★★☆|不要|適|Follow-up満たす|
# |NumPy使用|O(mn)|O(1)|低|★★☆|numpy|適|外部ライブラリ必要|

# ## 3. 採用アルゴリズムと根拠

# **選択**: 第一行・第一列をフラグとして使用するO(1)空間解法
# - **理由**: Follow-upのO(1)空間要求を満たし、Pythonのリスト操作に適合
# - **Python最適化戦略**:
#   - リスト内包表記の活用（該当する場面では）
#   - enumerate()による効率的なインデックス処理
#   - range()の最適化されたイテレーション
# - **トレードオフ**: 若干の実装複雑性 < 空間効率性の大幅改善

# ## 4. 実装パターン

"""
Set Matrix Zeroes - Python Implementation
CPython 3.11+ optimized solutions with type safety
"""

import time
from functools import wraps
from typing import Any, List, Tuple


class Solution:
    """
    Set Matrix Zeroes解決クラス

    競技プログラミング向けと業務開発向けの2パターンを提供
    """

    def setZeroes(self, matrix: List[List[int]]) -> None:
        """
        LeetCode標準インターフェース（競技プログラミング向け）

        Args:
            matrix: 2D integer matrix to modify in-place

        Time Complexity: O(mn)
        Space Complexity: O(1)
        """
        if not matrix or not matrix[0]:
            return

        m, n = len(matrix), len(matrix[0])

        # 第一行・第一列の元の状態チェック
        first_row_zero = any(matrix[0][j] == 0 for j in range(n))
        first_col_zero = any(matrix[i][0] == 0 for i in range(m))

        # 内部要素の0を検出してフラグ設定
        for i in range(1, m):
            for j in range(1, n):
                if matrix[i][j] == 0:
                    matrix[i][0] = 0  # 行フラグ
                    matrix[0][j] = 0  # 列フラグ

        # フラグに基づいて内部要素を0に設定
        for i in range(1, m):
            for j in range(1, n):
                if matrix[i][0] == 0 or matrix[0][j] == 0:
                    matrix[i][j] = 0

        # 第一行の処理
        if first_row_zero:
            for j in range(n):
                matrix[0][j] = 0

        # 第一列の処理
        if first_col_zero:
            for i in range(m):
                matrix[i][0] = 0

    def set_zeroes_production(self, matrix: List[List[int]]) -> None:
        """
        業務開発向け実装（型安全・エラーハンドリング重視）

        Args:
            matrix: 2D integer matrix to modify in-place

        Raises:
            TypeError: 入力型が不正な場合
            ValueError: 入力値が制約を満たさない場合

        Time Complexity: O(mn)
        Space Complexity: O(1)
        """
        # 入力検証
        self._validate_matrix_input(matrix)

        # エッジケース処理
        if self._is_empty_matrix(matrix):
            return

        dimensions = self._get_matrix_dimensions(matrix)
        m, n = dimensions.rows, dimensions.cols

        # メインアルゴリズム実行
        try:
            self._execute_zero_setting_algorithm(matrix, m, n)
        except Exception as e:
            raise RuntimeError(f"Algorithm execution failed: {e}") from e

    def _validate_matrix_input(self, matrix: Any) -> None:
        """型安全な入力検証"""
        if not isinstance(matrix, list):
            raise TypeError("Matrix must be a list")

        if not matrix:
            return  # 空行列は有効

        if not isinstance(matrix[0], list):
            raise TypeError("Matrix must be a 2D list")

        # 矩形チェック
        expected_cols = len(matrix[0])
        for i, row in enumerate(matrix):
            if not isinstance(row, list):
                raise TypeError(f"Row {i} must be a list")

            if len(row) != expected_cols:
                raise ValueError(
                    f"Matrix must be rectangular. Row {i} has {len(row)} elements, expected {expected_cols}"
                )

            # 要素の型チェック
            for j, element in enumerate(row):
                if not isinstance(element, int):
                    raise TypeError(
                        f"Element at [{i}][{j}] must be an integer, got {type(element)}"
                    )

        # サイズ制約チェック
        if len(matrix) > 200 or len(matrix[0]) > 200:
            raise ValueError("Matrix dimensions exceed maximum size (200x200)")

        # 値範囲チェック
        for i, row in enumerate(matrix):
            for j, element in enumerate(row):
                if element < -(2**31) or element >= 2**31:
                    raise ValueError(
                        f"Element at [{i}][{j}] out of 32-bit integer range"
                    )

    def _is_empty_matrix(self, matrix: List[List[int]]) -> bool:
        """空行列かどうかの判定"""
        return not matrix or not matrix[0]

    def _get_matrix_dimensions(self, matrix: List[List[int]]) -> "MatrixDimensions":
        """行列の次元を取得"""
        return MatrixDimensions(rows=len(matrix), cols=len(matrix[0]) if matrix else 0)

    def _execute_zero_setting_algorithm(
        self, matrix: List[List[int]], m: int, n: int
    ) -> None:
        """ゼロ設定アルゴリズムの実行"""
        # 第一行・第一列の元の状態を保存
        boundary_state = self._analyze_boundary_zeros(matrix, m, n)

        # 内部要素を走査してフラグを設定
        self._mark_zero_flags(matrix, m, n)

        # フラグに基づいて内部要素を更新
        self._apply_zero_flags(matrix, m, n)

        # 境界要素の処理
        self._handle_boundary_zeros(matrix, m, n, boundary_state)

    def _analyze_boundary_zeros(
        self, matrix: List[List[int]], m: int, n: int
    ) -> "BoundaryZeroState":
        """境界の0状態を分析"""
        first_row_has_zero = any(matrix[0][j] == 0 for j in range(n))
        first_col_has_zero = any(matrix[i][0] == 0 for i in range(m))

        return BoundaryZeroState(
            first_row_zero=first_row_has_zero, first_col_zero=first_col_has_zero
        )

    def _mark_zero_flags(self, matrix: List[List[int]], m: int, n: int) -> None:
        """0要素を検出してフラグをマーク"""
        for i in range(1, m):
            for j in range(1, n):
                if matrix[i][j] == 0:
                    matrix[i][0] = 0  # 行フラグ
                    matrix[0][j] = 0  # 列フラグ

    def _apply_zero_flags(self, matrix: List[List[int]], m: int, n: int) -> None:
        """フラグに基づいて内部要素を0に設定"""
        for i in range(1, m):
            for j in range(1, n):
                if matrix[i][0] == 0 or matrix[0][j] == 0:
                    matrix[i][j] = 0

    def _handle_boundary_zeros(
        self,
        matrix: List[List[int]],
        m: int,
        n: int,
        boundary_state: "BoundaryZeroState",
    ) -> None:
        """境界要素の0設定処理"""
        if boundary_state.first_row_zero:
            for j in range(n):
                matrix[0][j] = 0

        if boundary_state.first_col_zero:
            for i in range(m):
                matrix[i][0] = 0


# 型安全性のための補助クラス
class MatrixDimensions:
    """行列の次元情報"""

    __slots__ = ["rows", "cols"]

    def __init__(self, rows: int, cols: int) -> None:
        self.rows = rows
        self.cols = cols


class BoundaryZeroState:
    """境界の0状態情報"""

    __slots__ = ["first_row_zero", "first_col_zero"]

    def __init__(self, first_row_zero: bool, first_col_zero: bool) -> None:
        self.first_row_zero = first_row_zero
        self.first_col_zero = first_col_zero


# パフォーマンス測定デコレータ（開発時用）
def measure_performance(func):
    """実行時間測定デコレータ"""

    @wraps(func)
    def wrapper(*args, **kwargs):
        start_time = time.perf_counter()
        result = func(*args, **kwargs)
        end_time = time.perf_counter()
        print(f"{func.__name__} executed in {end_time - start_time:.6f} seconds")
        return result

    return wrapper


# テスト実行クラス
class TestRunner:
    """型安全なテスト実行"""

    def __init__(self) -> None:
        self.solution = Solution()

    def run_test_cases(self) -> None:
        """テストケースの実行"""
        test_cases = self._get_test_cases()

        for i, (input_matrix, expected, description) in enumerate(test_cases):
            self._run_single_test(i + 1, input_matrix, expected, description)

    def _get_test_cases(self) -> List[Tuple[List[List[int]], List[List[int]], str]]:
        """テストケースの定義"""
        return [
            (
                [[1, 1, 1], [1, 0, 1], [1, 1, 1]],
                [[1, 0, 1], [0, 0, 0], [1, 0, 1]],
                "Example 1",
            ),
            (
                [[0, 1, 2, 0], [3, 4, 5, 2], [1, 3, 1, 5]],
                [[0, 0, 0, 0], [0, 4, 5, 0], [0, 3, 1, 0]],
                "Example 2",
            ),
            ([[1]], [[1]], "Single element - no zero"),
            ([[0]], [[0]], "Single element - zero"),
            ([[0, 0], [0, 0]], [[0, 0], [0, 0]], "All zeros"),
            (
                [[1, 2, 3], [4, 0, 6], [7, 8, 9]],
                [[1, 0, 3], [0, 0, 0], [7, 0, 9]],
                "Center zero",
            ),
        ]

    def _run_single_test(
        self,
        test_num: int,
        input_matrix: List[List[int]],
        expected: List[List[int]],
        description: str,
    ) -> None:
        """単一テストケースの実行"""
        # 深いコピーを作成（元データ保護）
        test_matrix = [row[:] for row in input_matrix]

        try:
            # 競技プログラミング版でテスト
            self.solution.setZeroes(test_matrix)

            if self._matrices_equal(test_matrix, expected):
                print(f"Test {test_num} ({description}): PASS")
            else:
                print(f"Test {test_num} ({description}): FAIL")
                print(f"  Expected: {expected}")
                print(f"  Got:      {test_matrix}")

        except Exception as e:
            print(f"Test {test_num} ({description}): ERROR - {e}")

    def _matrices_equal(self, a: List[List[int]], b: List[List[int]]) -> bool:
        """行列の等価性チェック"""
        if len(a) != len(b):
            return False

        for i in range(len(a)):
            if len(a[i]) != len(b[i]):
                return False
            for j in range(len(a[i])):
                if a[i][j] != b[i][j]:
                    return False

        return True


# 使用例
def main() -> None:
    """メイン実行関数"""
    print("=== Set Matrix Zeroes - Python Implementation ===")

    # テスト実行
    test_runner = TestRunner()
    test_runner.run_test_cases()

    # パフォーマンステスト用大規模行列生成
    def generate_test_matrix(
        rows: int, cols: int, zero_rate: float = 0.1
    ) -> List[List[int]]:
        """テスト用行列生成"""
        import random

        matrix = [[1 for _ in range(cols)] for _ in range(rows)]
        zero_count = int(rows * cols * zero_rate)

        for _ in range(zero_count):
            i = random.randint(0, rows - 1)
            j = random.randint(0, cols - 1)
            matrix[i][j] = 0

        return matrix

    # 大規模テスト例（コメントアウト）
    # large_matrix = generate_test_matrix(200, 200, 0.05)
    # solution = Solution()
    # solution.setZeroes(large_matrix)


if __name__ == "__main__":
    main()

# ## 5. Python特有最適化ポイント

# ### CPython インタープリター最適化
# - **組み込み関数活用**: `any()`, `range()`, `len()` の効率的利用
# - **リスト内包表記**: 条件チェックでの `any()` 活用
# - **イテレータ最適化**: `range()` による効率的なループ処理
# - **メモリ効率**: `__slots__` による小規模クラスの最適化

# ### データ構造選択指針
# - **list**: in-place操作に最適、インデックスアクセスが高速
# - **スライシング回避**: `matrix[0][j]` vs `matrix[0][j:j+1]` の選択
# - **メモリ局所性**: 行優先アクセスパターンでのキャッシュ効率

# ### メモリ最適化
# - **in-place操作**: 追加メモリ割り当て回避
# - **浅いコピー**: テスト時の `row[:]` による効率的コピー
# - **プリミティブ型**: int型の直接操作でオーバーヘッド最小化

# ## 6. CPython 3.11+ 新機能活用

# ### 型アノテーション改良
# - **厳密な型ヒント**: `List[List[int]]` による2次元配列の型安全性
# - **TypeGuard**: 入力検証での型絞り込み
# - **Generic使用**: 再利用可能な型安全コンポーネント

# ### パフォーマンス改善
# - **例外処理最適化**: try-except文のオーバーヘッド削減
# - **関数呼び出し高速化**: 小関数のインライン化促進

# ## 7. パフォーマンス考察

# ### 理論計算量
# - **時間計算量**: O(mn) - 各要素への定数回アクセス
# - **空間計算量**: O(1) - 固定数の変数のみ使用

# ### CPython実測予想
# - **組み込み関数**: C実装による高速な `any()`, `range()` 処理
# - **リストアクセス**: 連続メモリアクセスによるキャッシュ効率
# - **型チェック**: 開発時のみで実行時オーバーヘッドなし

# ### 改善余地
# - **NumPy活用**: 大規模行列では NumPy のベクトル化操作
# - **Cython導入**: CPU集約的部分のC拡張
# - **並列処理**: multiprocessing による行並列処理（GIL回避）

# ## 8. 業務開発 vs 競技プログラミング

# ### 業務開発版の特徴
# - **完全な型安全性**: mypy/pylance対応の厳密な型ヒント
# - **包括的エラーハンドリング**: 想定される全例外の適切な処理
# - **自己文書化**: docstring、型ヒント、明確な変数名
# - **テスタビリティ**: 各機能の独立性と単体テスト容易性

# ### 競技プログラミング版の特徴
# - **最小限実装**: 必要最低限のコードで最大性能
# - **エラーチェック省略**: 制約条件を信頼した最適化
# - **直接的アルゴリズム**: 抽象化を排した効率重視の実装

# この実装はPythonの特性を最大限活用しつつ、型安全性と実行効率を両立させ、Follow-upの要求も完全に満たしています。CPython 3.11+の最適化と標準ライブラリの効率的な活用により、実用的なパフォーマンスを実現しています。

File: Algorithm/Other/leetcode/73. Set Matrix Zeroes/Claude/Set-Matrix-Zeroes.ts
// ## 1. 問題分析結果

// ### 競技プログラミング視点

// - **制約**: m×n 行列、1≤m,n≤200 → 最大 40,000 要素
// - **要求**: in-place 操作で O(1)空間計算量が理想
// - **キーポイント**: 型安全性を保ちながら最適なメモリ使用量を実現

// ### 業務開発視点

// - **型安全性**: コンパイル時に境界条件エラーを検出
// - **保守性**: インターフェース定義による明確な契約
// - **エラーハンドリング**: 型ガードによる実行時安全性確保

// ### TypeScript 特有分析

// - **型システム活用**: 2 次元配列の型安全な操作
// - **コンパイル時最適化**: `readonly`修飾子による意図明確化
// - **V8 最適化との相乗効果**: 型情報による JIT コンパイル効率化

// ## 2. アルゴリズム比較表

// | アプローチ   | 時間計算量 | 空間計算量 | TS 実装コスト | 可読性 | 型安全性 | V8 最適化 | 備考                 |
// | ------------ | ---------- | ---------- | ------------- | ------ | -------- | --------- | -------------------- |
// | 追加配列使用 | O(mn)      | O(m+n)     | 低            | ★★★    | 高       | 適        | 型定義が直感的       |
// | 第一行列使用 | O(mn)      | O(1)       | 中            | ★★☆    | 高       | 適        | Follow-up 要求満たす |
// | ビット操作   | O(mn)      | O(1)       | 高            | ★☆☆    | 中       | 不適      | 型安全性に制約       |

// ## 3. 採用アルゴリズムと根拠

// **選択**: 型安全な第一行・第一列フラグ使用の O(1)空間解法

// - **理由**: TypeScript の型システムで安全性を担保しつつ、Follow-up の要求を満たす
// - **TypeScript 最適化戦略**:
//   - 厳密な型定義による境界エラー防止
//   - `readonly`配列による意図の明確化
//   - 型ガードによる実行時安全性確保
// - **トレードオフ**: 型注釈のオーバーヘッド < コンパイル時エラー検出の利益

// ## 4. 実装詳細

// ```python
// 型定義
type Matrix = number[][];
type ReadonlyMatrix = readonly (readonly number[])[];

interface MatrixDimensions {
    readonly rows: number;
    readonly cols: number;
}

interface MatrixValidation {
    readonly isValid: boolean;
    readonly error?: string;
}

/**
 * Set Matrix Zeroes - TypeScript O(1) Space Solution
 * @param matrix - 2D integer matrix to modify in-place
 * @complexity Time: O(mn), Space: O(1)
 * @throws {TypeError} Invalid matrix input
 * @throws {RangeError} Matrix dimensions out of bounds
 */
function setZeroes(matrix: Matrix): void {
    // 入力検証（型安全）
    const validation = validateMatrix(matrix);
    if (!validation.isValid) {
        throw new TypeError(validation.error || 'Invalid matrix input');
    }

    const dimensions = getMatrixDimensions(matrix);
    const { rows: m, cols: n } = dimensions;

    // エッジケース: 空行列（TypeScriptでは型レベルで防げるが実行時チェック）
    if (m === 0 || n === 0) {
        return;
    }

    // 第一行・第一列の元の状態を型安全に保存
    const firstRowHasZero = hasZeroInFirstRow(matrix, n);
    const firstColHasZero = hasZeroInFirstCol(matrix, m);

    // フラグマーキング（型安全な範囲チェック付き）
    markZeroFlags(matrix, m, n);

    // 内部要素のゼロ化（型ガード付き）
    applyZeroFlags(matrix, m, n);

    // 境界処理（型安全）
    if (firstRowHasZero) {
        setFirstRowZero(matrix, n);
    }

    if (firstColHasZero) {
        setFirstColZero(matrix, m);
    }
}

/**
 * 型安全な行列検証
 */
function validateMatrix(matrix: unknown): MatrixValidation {
    // null/undefined チェック
    if (matrix == null) {
        return { isValid: false, error: 'Matrix cannot be null or undefined' };
    }

    // 配列型チェック
    if (!Array.isArray(matrix)) {
        return { isValid: false, error: 'Input must be an array' };
    }

    // 空配列チェック
    if (matrix.length === 0) {
        return { isValid: true }; // 有効なエッジケース
    }

    // 各行の検証
    for (let i = 0; i < matrix.length; i++) {
        const row = matrix[i];

        if (!Array.isArray(row)) {
            return { isValid: false, error: `Row ${i} is not an array` };
        }

        // 矩形チェック
        if (i > 0 && row.length !== matrix[0].length) {
            return { isValid: false, error: 'Matrix must be rectangular' };
        }

        // 各要素の型チェック
        for (let j = 0; j < row.length; j++) {
            if (typeof row[j] !== 'number') {
                return {
                    isValid: false,
                    error: `Element at [${i}][${j}] is not a number`,
                };
            }
        }
    }

    return { isValid: true };
}

/**
 * 行列次元の型安全な取得
 */
function getMatrixDimensions(matrix: Matrix): MatrixDimensions {
    return {
        rows: matrix.length,
        cols: matrix.length > 0 ? matrix[0].length : 0,
    } as const;
}

/**
 * 第一行のゼロ存在チェック（型安全）
 */
function hasZeroInFirstRow(matrix: ReadonlyMatrix, cols: number): boolean {
    for (let j = 0; j < cols; j++) {
        if (matrix[0][j] === 0) {
            return true;
        }
    }
    return false;
}

/**
 * 第一列のゼロ存在チェック（型安全）
 */
function hasZeroInFirstCol(matrix: ReadonlyMatrix, rows: number): boolean {
    for (let i = 0; i < rows; i++) {
        if (matrix[i][0] === 0) {
            return true;
        }
    }
    return false;
}

/**
 * ゼロフラグのマーキング（V8最適化考慮）
 */
function markZeroFlags(matrix: Matrix, rows: number, cols: number): void {
    // V8最適化: インデックスによる連続アクセス
    for (let i = 1; i < rows; i++) {
        for (let j = 1; j < cols; j++) {
            if (matrix[i][j] === 0) {
                matrix[i][0] = 0; // 行フラグ
                matrix[0][j] = 0; // 列フラグ
            }
        }
    }
}

/**
 * フラグに基づくゼロ適用（型安全）
 */
function applyZeroFlags(matrix: Matrix, rows: number, cols: number): void {
    // 型安全な境界チェック付きループ
    for (let i = 1; i < rows; i++) {
        for (let j = 1; j < cols; j++) {
            if (matrix[i][0] === 0 || matrix[0][j] === 0) {
                matrix[i][j] = 0;
            }
        }
    }
}

/**
 * 第一行のゼロ化（型安全）
 */
function setFirstRowZero(matrix: Matrix, cols: number): void {
    for (let j = 0; j < cols; j++) {
        matrix[0][j] = 0;
    }
}

/**
 * 第一列のゼロ化（型安全）
 */
function setFirstColZero(matrix: Matrix, rows: number): void {
    for (let i = 0; i < rows; i++) {
        matrix[i][0] = 0;
    }
}

// ユーティリティ型定義
type TestCase = {
    readonly input: Matrix;
    readonly expected: Matrix;
    readonly description: string;
};

/**
 * 型安全なテストケース実行
 */
function runTestCase(testCase: TestCase): void {
    const { input, expected, description } = testCase;

    // 入力の深いコピー（型安全）
    const testMatrix: Matrix = input.map((row) => [...row]);

    try {
        setZeroes(testMatrix);

        // 結果検証（型安全な比較）
        const isEqual = matricesEqual(testMatrix, expected);
        console.log(`${description}: ${isEqual ? 'PASS' : 'FAIL'}`);

        if (!isEqual) {
            console.log('  Expected:', JSON.stringify(expected));
            console.log('  Actual:  ', JSON.stringify(testMatrix));
        }
    } catch (error) {
        console.log(
            `${description}: ERROR - ${error instanceof Error ? error.message : 'Unknown error'}`,
        );
    }
}

/**
 * 型安全な行列比較
 */
function matricesEqual(a: ReadonlyMatrix, b: ReadonlyMatrix): boolean {
    if (a.length !== b.length) return false;

    for (let i = 0; i < a.length; i++) {
        if (a[i].length !== b[i].length) return false;

        for (let j = 0; j < a[i].length; j++) {
            if (a[i][j] !== b[i][j]) return false;
        }
    }

    return true;
}

// 使用例とテストケース
function testSetZeroesTypeScript(): void {
    const testCases: readonly TestCase[] = [
        {
            input: [
                [1, 1, 1],
                [1, 0, 1],
                [1, 1, 1],
            ],
            expected: [
                [1, 0, 1],
                [0, 0, 0],
                [1, 0, 1],
            ],
            description: 'Example 1',
        },
        {
            input: [
                [0, 1, 2, 0],
                [3, 4, 5, 2],
                [1, 3, 1, 5],
            ],
            expected: [
                [0, 0, 0, 0],
                [0, 4, 5, 0],
                [0, 3, 1, 0],
            ],
            description: 'Example 2',
        },
        {
            input: [[1]],
            expected: [[1]],
            description: 'Single element - no zero',
        },
        {
            input: [[0]],
            expected: [[0]],
            description: 'Single element - zero',
        },
        {
            input: [
                [0, 0],
                [0, 0],
            ],
            expected: [
                [0, 0],
                [0, 0],
            ],
            description: 'All zeros',
        },
    ] as const;

    testCases.forEach(runTestCase);
}

// パフォーマンス測定用（開発時用）
function generateTypedTestMatrix(rows: number, cols: number, zeroRate = 0.1): Matrix {
    // 型安全な行列生成
    const matrix: Matrix = Array(rows)
        .fill(null)
        .map(() => Array(cols).fill(1));
    const zeroCount = Math.floor(rows * cols * zeroRate);

    for (let k = 0; k < zeroCount; k++) {
        const i = Math.floor(Math.random() * rows);
        const j = Math.floor(Math.random() * cols);
        matrix[i][j] = 0;
    }

    return matrix;
}

// エクスポート（モジュール使用時）
export { setZeroes, validateMatrix, getMatrixDimensions, testSetZeroesTypeScript };
export type { Matrix, ReadonlyMatrix, MatrixDimensions, TestCase };
// ```

// ## 5. TypeScript 特有最適化ポイント

// ### 型システム活用

// - **厳密な型定義**: `Matrix`型による 2 次元配列の型安全性確保
// - **型ガード実装**: `validateMatrix`による実行時型チェック
// - **`readonly`修飾子**: 不変性の意図明確化とコンパイラ最適化
// - **Union Types**: エラー処理での型安全な結果表現

// ### V8 エンジン最適化（JavaScript 同等）

// - **インライン化促進**: 小関数に分割した処理ロジック
// - **Hidden Class 安定化**: 一貫した配列アクセスパターン
// - **効率的ループ**: for ループによるインデックスアクセス
// - **局所性活用**: 連続メモリアクセス最適化

// ### TypeScript 特有の最適化

// - **`as const`アサーション**: テストケースの型推論最適化
// - **型注釈最小化**: 必要箇所のみの明示的型指定
// - **Template Literal Types**: 将来的な拡張性考慮
// - **Utility Types 活用**: `readonly`での不変性保証

// ## 6. パフォーマンス考察

// ### 理論計算量

// - **時間計算量**: O(mn) - 各要素への定数回アクセス
// - **空間計算量**: O(1) - 固定数の変数のみ使用

// ### TypeScript/V8 実測予想

// - **コンパイル後 JavaScript**: 型情報除去により実行時オーバーヘッドなし
// - **JIT 最適化**: 型情報によるより効率的な機械語生成
// - **メモリ効率**: 型チェックはコンパイル時のみで実行時影響なし
// - **開発効率**: コンパイル時エラー検出による品質向上

// ### 型システムオーバーヘッド

// - **コンパイル時**: 型チェック処理によるビルド時間若干増加
// - **実行時**: ゼロオーバーヘッド（型情報は削除される）
// - **バンドルサイズ**: TypeScript メタデータなしで同等サイズ

// ### 改善余地

// - **Generic Types**: より柔軟な型パラメータ化
// - **Brand Types**: より厳密な型安全性（行列次元の型レベル保証）
// - **Template Literal Types**: 動的な型生成での更なる安全性
// - **Conditional Types**: 条件に基づく型分岐による最適化

// ## 7. TypeScript 導入効果

// ### 開発時利益

// - **コンパイル時エラー検出**: 境界条件違反の事前発見
// - **IDE 支援強化**: 自動補完・リファクタリング支援
// - **ドキュメント効果**: 型定義による自己文書化
// - **リファクタリング安全性**: 型チェックによる変更影響範囲把握

// ### 保守性向上

// - **契約明確化**: インターフェース定義による仕様明示
// - **変更検出**: 型システムによる破壊的変更の自動検出
// - **テスト品質**: 型安全なテストケース記述

// この実装は TypeScript の型システムを最大限活用しながら、JavaScript と同等の実行時パフォーマンスを実現し、Follow-up の要求も完全に満たしています。
